go/vt/vttablet/tabletmanager/rpc_replication.go

```diff
@@ -421,12 +421,28 @@ func (agent *ActionAgent) demoteMaster(ctx context.Context, revertPartialFailure
 		}
 	}()
 
-	// Return the current replication position.
-	pos, err := agent.MysqlDaemon.MasterPosition()
+	// if a tablet was accepting writes, it cannot be replicating from another server,
+	// so we return the master position.
+	// We might call DemoteMaster on replicas in a no-master situation, and use the result to choose a master.
+	// So for replicas, we return the relay log position instead.
+	if !wasReadOnly {
+		// Return the current master position.
+		pos, err := agent.MysqlDaemon.MasterPosition()
+		if err != nil {
+			return "", err
+		}
+		return mysql.EncodePosition(pos), nil
+	}
+	rs, err := agent.MysqlDaemon.SlaveStatus()
 	if err != nil {
-		return "", err
+		// fall back to masterPosition
+		pos, err := agent.MysqlDaemon.MasterPosition()
+		if err != nil {
+			return "", err
+		}
+		return mysql.EncodePosition(pos), nil
 	}
-	return mysql.EncodePosition(pos), nil
+	return mysql.EncodePosition(rs.RelayLogPosition), nil
```

Comment:
Currently, if there are errant transactions on this replica, they'll show up in the returned position. Does that end up protecting us against any scenario? Do we lose that if we return the relay log pos instead?

Maybe we should check for this case and fail if our expected invariant does not hold? If one of relayPos or masterPos are a superset of the other (`AtLeast()`), then return the one that's farther ahead. If neither is a superset of the other, return an error saying we detected errant transactions and this replica should not be used as a master.

Reply:
Actually, if masterPos is a superset of relayPos, that is exactly the errant transaction situation. The latest commit addresses this.

Reply:
Based on experiments with relayLogPosition, we have decided to exclude that logic from this PR, and do it as a separate PR.